src/torch_models.py

```diff
@@ -319,13 +319,32 @@ def _create_loss_fn(self) -> Callable[[Tensor, Tensor], Tensor]:
 
 
 class ImplicitMLPRegressor(PyTorchRegressor):
-    """A regressor implemented via an "energy function".
+    """A regressor implemented via an energy function.
 
-    Currently the energy function is implemented as a binary classifier, which
-    is not consistent with how energy functions are usually treated/trained.
-    This will change soon.
+    For each positive (x, y) pair, a number of "negative" (x, y') pairs are
+    generated. The model is then trained to distinguish positive from negative
+    conditioned on x using a contrastive loss.
 
-    Negative examples are generated within the class.
+    The implementation idea is the following. We want to use a contrastive
+    loss that looks like this:
+
+        L = E[-log(p(y | x, {y'}))]
+
+        p(y | x, {y'})) = exp(-f(x, y)) / [
+            (exp(-f(x, y)) + sum_{y'} exp(-f(x, y')))
+        ]
+
+    where (x, y) is an example "positive" input/output from (X, Y), f is
+    the model that we are learning in this class, and {y'} is a set of
+    "negative" output examples for input x. The size of that set is
+    self._num_negatives_per_input (num_negatives).
```

Comment:
what does the `(num_negatives)` mean here -- i thought this sentence is sufficiently clear without it